File: 695. Max Area of Island/695. Max Area of Island.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
var maxAreaOfIsland = function (grid) {
  const m = grid.length;
  const n = grid[0].length;
  let maxArea = 0;

  const dfs = (i, j) => {
    if (i < 0 || i >= m || j < 0 || j >= n || grid[i][j] !== 1) {
      return 0;
    }

    grid[i][j] = 0;

    return 1 + dfs(i + 1, j) + dfs(i - 1, j) + dfs(i, j + 1) + dfs(i, j - 1);
  };

  for (let i = 0; i < m; i++) {
    for (let j = 0; j < n; j++) {
      if (grid[i][j] === 1) {
        const currentArea = dfs(i, j);
        maxArea = Math.max(maxArea, currentArea);
      }
    }
  }

  return maxArea;
};

File: 695. Max Area of Island.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        ROWS, COLS = len(grid), len(grid[0])

        max_area = 0

        for r in range(ROWS):
            for c in range(COLS):
                if grid[r][c] == 1:
                    grid[r][c] = 0
                    queue = deque([(r, c)])
                    area = 0
                    while queue:
                        area += 1
                        cr, cc = queue.popleft()
                        for dr, dc in [(1, 0), (-1, 0), (0, 1), (0, -1)]:
                            nr, nc = cr + dr, cc + dc
                            if 0 <= nr < ROWS and 0 <= nc < COLS and grid[nr][nc] == 1:
                                grid[nr][nc] = 0
                                queue.append((nr, nc))
                    max_area = max(max_area, area)

        return max_area
